### PR TITLE
🐛 differentiate machine pool nodes with similar providerIDs

### PR DESCRIPTION
**Pull request by alexeldeib**
**What this PR does / why we need it**:
Provider ID comparisons for machine pools work a bit poorly right now. 1 Azure VMSS (read: ASG) corresponds to 1 CAPI MachinePool. Within a VMSS, instance IDs are numeric and unique. Between multiple VMSS, instance IDs are reused. When I say instance ID, I'm overloading the term: VMSS literally uses numeric IDs (0, 1, etc), and CAPI also parses that number as the "instance" here. So when we check for equality using cloudprovider == azure and instance ids are the same, that is a **_very_** broad match, and is not accurate.

The current logic in machinepool node ref controller (below) lists **all** nodes, and then compares them using the provider ID package to the list of providerIDs from a specific machine pool. 

Because instance IDs are not unique across VMSS, and the noderef controller doesn't filter nodes first to only those in the machine pools, the logic matches node refs to VMs from the wrong VMSS. 

This PR changes the node ref controller behavior to filter by annotations (really should be labels, why aren't we applying those for machinepools?) to only include nodes with an annotation added from the owner machine pool.

https://github.com/kubernetes-sigs/cluster-api/blob/37c862b5c9256d7ae65cd8ee8ddd7a103d51fe90/exp/controllers/machinepool_controller_noderef.go#L179-L221

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1503

### CODE REVIEW

File: exp/controllers/machinepool_controller_noderef.go [hunk 1]
```diff
@@ -173,53 +173,58 @@ func (r *MachinePoolReconciler) deleteRetiredNodes(ctx context.Context, c client
...
 	nodeList := corev1.NodeList{}
 	for {
 		if err := c.List(ctx, &nodeList, client.Continue(nodeList.Continue)); err != nil {
 			return getNodeReferencesResult{}, errors.Wrapf(err, "failed to List nodes")
 		}
 
-		for _, node := range nodeList.Items {
+		for idx := range nodeList.Items {
+			node := nodeList.Items[idx]
+			if node.ObjectMeta.Annotations[clusterv1.OwnerKindAnnotation] != "MachinePool" || node.ObjectMeta.Annotations[clusterv1.OwnerNameAnnotation] != owner {
```
Comment (LochanRn) [inline exp/controllers/machinepool_controller_noderef.go:199]: [MachinePool](https://github.com/kubernetes-sigs/cluster-api/blob/c6380c53e27a927a7cf22fd43b43c357a57925a7/exp/controllers/machinepool_controller_noderef.go#L117) annotations for the nodes are set after we fetch the nodeRefs.

Comment (alexeldeib) [reply exp/controllers/machinepool_controller_noderef.go:199]: hmm…might need to wait for the full provider ID fix in that case…

Comment (LochanRn) [reply exp/controllers/machinepool_controller_noderef.go:199]: Yeah i was thinking the same. In the current code why are we not using the full provider id by the way ?

Comment (alexeldeib) [reply exp/controllers/machinepool_controller_noderef.go:199]: I don't think there's a good reason, see https://github.com/kubernetes-sigs/cluster-api/pull/4913#issuecomment-878041148

I think we just need to do the work to fix/normalize that everywhere

**Comment by alexeldeib** [comment]
/close

guess this doesn't work well either :/ we'll just have to do the full provider ID comparison as suggested over in https://github.com/kubernetes-sigs/cluster-api/pull/4913#issuecomment-878041148

**Pull request closed without merging**